### PR TITLE
Add `InMemoryTraceClientWithTracking` to log traces

### PR DESCRIPTION
**Pull request by harupy**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/harupy/mlflow/pull/11618?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11618/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11618
```

</p>
</details>

### Related Issues/PRs

 #xxx

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

Run the following code with the databricks backend:

```python
import time
import uuid

import mlflow
from mlflow.entities.trace import Trace
from mlflow.entities.trace_data import TraceData
from mlflow.entities.trace_info import TraceInfo
from mlflow.entities.trace_status import TraceStatus
from mlflow.tracing.clients import get_trace_client

mlflow.set_tracking_uri("databricks")
name = "/Users/harutaka.kawamura@databricks.com/log-traces"
mlflow.set_experiment(name)
exp = mlflow.get_experiment_by_name(name)

tag = str(uuid.uuid4())

trace = Trace(
    info=TraceInfo(
        request_id="a",
        experiment_id=exp.experiment_id,
        timestamp_ms=int(time.time() * 1000),
        execution_time_ms=1000,
        status=TraceStatus.OK,
        tags={"test": tag},
    ),
    data=TraceData(),
)
trace_client = get_trace_client()
trace_client.log_trace(trace)

client = mlflow.MlflowClient()

traces = client.search_traces(experiment_ids=[exp.experiment_id], max_results=1)
print(traces)
assert "test" in traces[0].info.tags
assert traces[0].info.tags["test"] == tag

trace = client.get_trace(traces[0].info.request_id)
print(trace)


@mlflow.trace
def f():
    pass


@mlflow.trace
def g(x, y):
    return x + y


t = int(time.time() * 1000)
f()
g(3, 4)


with mlflow.start_span(name="test"):
    pass

traces = client.search_traces(
    experiment_ids=[exp.experiment_id],
    filter_string=f"timestamp_ms >= {t}",
    max_results=10,
)
assert len(traces) == 3
print(traces)
```

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/tracing/clients/local.py [hunk 1]
```diff
@@ -97,3 +98,15 @@ def get_trace(self, request_id: str) -> Optional[Trace]:
...
+class InMemoryTraceClientWithTracking(InMemoryTraceClient):
+    """
+    `InMemoryTraceClient` with tracking capabilities.
+    """
+
+    def log_trace(self, trace: Trace):
+        super().log_trace(trace)
+        client = MlflowClient()
+        client._create_trace_info(trace.trace_info)
+        client._upload_trace_data(trace.trace_info, trace.trace_data)
```
Comment (harupy) [inline mlflow/tracing/clients/local.py:112]: https://github.com/mlflow/mlflow/pull/11646

File: mlflow/tracing/clients/__init__.py
Comment (harupy) [inline mlflow/tracing/clients/__init__.py:1]: I need to add tests

**Comment by harupy** [comment]
I need to add tests + do manual testing.

**Comment by dbczumar** [comment]
@harupy Looks good! Can we add a thorough manual test plan to the PR description and also add unit tests?

**Review by B-Step62 — APPROVED** [review]
LGTM!

@dbczumar we discussed offline and decided to merge this first, then replace `_create_trace_info()` with `end_trace()` in the follow-up of #11734 to avoid parallel work confusion.

File: mlflow/tracing/clients/local.py [hunk 2]
```diff
@@ -66,3 +66,27 @@ def get_trace(self, request_id: str) -> Optional[Trace]:
 
     def _flush(self):
         self.queue.clear()
+
+
+class InMemoryTraceClientWithTracking(InMemoryTraceClient):
```
Comment (B-Step62) [inline mlflow/tracing/clients/local.py:71]: nit, not a blocker: Maybe we wanna have this in other file like `clients/databricks.py` as it's not longer local.

Comment (harupy) [reply mlflow/tracing/clients/local.py:71]: hmm maybe tracking.py?

Comment (harupy) [reply mlflow/tracing/clients/local.py:71]: or remote.py

**Pull request merged**